src/Microsoft.AspNetCore.Routing/Internal/ArrayBuilder.cs

```diff
@@ -0,0 +1,169 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// LICENSING NOTE: This file is from the dotnet corefx repository.
+// 
+// See https://github.com/dotnet/corefx/blob/143df51926f2ad397fef9c9ca7ede88e2721e801/src/Common/src/System/Collections/Generic/ArrayBuilder.cs 
+
+
+using System;
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.Routing.Internal
+{
+    /// <summary>
+    /// Helper type for avoiding allocations while building arrays.
+    /// </summary>
+    /// <typeparam name="T">The element type.</typeparam>
+    internal struct ArrayBuilder<T>
```

Comment:
This builder saves an allocation from the list. Does it do anything else?

Reply:
Due to the fact that it is only used for constructing an array, it does not need to perform any version checks and in general it perform less argument checks (eg. doesn't throw if negative capacity is specified). In other words, it uses less memory and and provides better throughput.


If the **TemplateBinder** ctor is not performance critical then we could just use **List\<T>**, but that offers no benefit.

I've updated the PR description to show results for both **List\<T>** and **ArrayBuilder\<T>**.